File: xortris/submissions/accepted/vinyals.hs
import Data.List (transpose)

flip1 '.' = 'X'
flip1 'X' = '.'
flipfirst3 board = map (\x -> (if snd x < 3 then flip1 else id) (fst x)) (zip board [0..])

solve1dim board | length board <= 3 = all (=='.') board
solve1dim ('.':board) = solve1dim board
solve1dim ('X':board) = solve1dim (flipfirst3 board)

countx str = length (filter (=='X') str)

solvemn 1 n strl = solve1dim (concat strl)
solvemn m 1 strl = solve1dim (concat strl)
solvemn 2 2 strl = countx (concat strl) `mod` 4 == 0
solvemn m n strl = countx (concat strl) `mod` 2 == 0

solve strl =
  let [m,n] = [read token :: Int | token <- words (head strl)]
  in solvemn m n (tail strl)

moin str = if solve (lines str) then "possible\n" else "impossible\n"
main = interact moin

File: xortris/submissions/accepted/vinyals.cc
#include <iostream>
using namespace std;

bool solve1(__int128 board, int n) {
  if (n<=3) return board==0;
  if (board&1) return solve1((board^0xf)>>1,n-1);
  return solve1(board>>1,n-1);
}

bool solve1(const string& board) {
  int n=board.size();
  __int128 boardint=0;
  for (char c:board) {
    boardint<<=1;
    if (c=='X') boardint|=1;
  }
  return solve1(boardint,n);
}

bool solve() {
  int m,n;
  cin >> m >> n;
  string board;
  if (m==1) {
    cin >> board;
    return solve1(board);
  }
  else if (n==1) {
    board.resize(m);
    for (int i=0;i<m;++i) cin >> board[i];
    return solve1(board);
  }
  else {
    int count=0;
    for (int i=0;i<m;++i) {
      for (int j=0;j<n;++j) {
        char c;
        cin >> c;
        count += (c=='X'?1:0);
      }
    }
    if (n==2 and m==2) return !(count&3);
    return !(count&1);
  }
}

int main() {
  cout << (solve()?"possible":"impossible") << endl;
}
